velox/exec/Window.cpp

```diff
@@ -350,10 +388,50 @@ void Window::callApplyForPartitionRows(
     rawPeerEnds[j] = peerEndRow_ - 1 - firstPartitionRow;
   }
 
-  auto updateFrameBounds = [&](vector_size_t* rawFrameBounds,
+  auto readFrameValues = [&](const FrameChannelArg& frameArg) -> void {
+    frameArg.value->resize(numRows);
+    windowPartition_->extractColumn(
+        frameArg.index, partitionOffset_, numRows, 0, frameArg.value);
+  };
+
+  auto columnValuesCheck = [&](const int64_t* offsets) -> void {
+    if (!std::all_of(offsets, offsets + numRows, [&](vector_size_t i) {
+          return i > 0;
+        })) {
+      VELOX_FAIL("k in frame bounds must be at least 1");
+    }
+  };
+
+  auto updateKRowsFrameBounds = [&](vector_size_t* rawFrameBounds,
+                                    const bool isKPreceding,
+                                    const FrameChannelArg& frameArg) -> void {
+    if (frameArg.index == kConstantChannel) {
+      vector_size_t constantOffset =
+          frameArg.value->template as<ConstantVector<int64_t>>()->valueAt(0);
+      std::iota(
+          rawFrameBounds,
+          rawFrameBounds + numRows,
+          startRow + (isKPreceding ? -constantOffset : constantOffset) -
+              firstPartitionRow);
+    } else {
+      readFrameValues(frameArg);
+      auto offsetsVector = frameArg.value->asFlatVector<int64_t>();
+      columnValuesCheck(offsetsVector->rawValues());
```

Comment:
Since we only need to read values here, we can simply get values() from the original VectorPtr (without a conversion to FlatVector). values() returns a BufferPtr and we can just use as() to get the int64* for it.

Reply:
Thanks for the suggestion, updated accordingly.